Attributes use enums instead of strings as keys

vintagedave opened the pull request:
A syntax node's Attributes[] dictionary (and related methods) now use an enum for the attribute name instead of a string.

This is unlikely to reduce string memory usage much, since the compiler merges identical compile-time strings, but is a plus for correctness (compiler catches spelling errors) and is neater.  It is also a good basis for upcoming code to merge all attribute values - coming soon :)

RomanYankovsky commented:
Looks good, I'll merge this soon.

By the way, did you do any tests? Does it consume less memory now?

vintagedave commented:
Great :)

Not yet.  I doubt it will reduce memory usage much, since I think most keys
are likely to be hardcoded at compile time, eg writing Attributes['foo'],
so folded together by the compiler.  Whereas the attribute values are
generated at runtime, probably with many duplicates.  I'm working on a
patch for reducing runtime string usage right now.

IMO the major benefit of enums here is just compile-time checking - it
avoids writing Attributes['neme'] for example.  That wouldn't cause an
error before, just a runtime bug, whereas now the equivalent,
Attributes[anNeme], is a compile-time error.


On 4 August 2015 at 17:39, Roman Yankovsky <notifications@github.com> wrote:

> Looks good, I'll merge this soon.
>
> By the way, did you do any tests? Does it consume less memory now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RomanYankovsky/DelphiAST/pull/112#issuecomment-127653392>
> .
>

vintagedave commented:
Hi Roman - I've updated this. Please look at TPasSyntaxTreeBuilder.DirectiveBinding. That method used the lexer token as an attribute name, which obviously wouldn't work as an enum, and in the previous commit I replaced it with an assert and forgot about it.  The current code status has this fixed and has a few new enums.

RomanYankovsky commented on Source/DelphiAST.Consts.pas:
```diff
@@ -254,15 +269,32 @@ interface
+    'abstract'
+  );
+begin
+  Exit(AttributeNameStrings[AttributeName]);
```
People will complain that this does not compile in older Delphi versions :)

NickRing replied:
But hasn't `Exit(<value)` been there since `TDictionary` existed?

RomanYankovsky replied:
Hm, perhaps. Then ignore my comment :)

NickRing replied:
Going by http://www.tindex.net/Language/Delphi2009.html both came in to existence at the same time. Q: Does FPC have `Exit(<value>)`?

Wosi replied:
A: Yes, FPC has the Exit command.
